ocrd_network/ocrd_network/processing_server.py

```diff
@@ -266,16 +289,76 @@ def query_ocrd_tool_json_from_server(self, processor_name):
         return ocrd_tool, processor_server_url
 
     async def push_processor_job(self, processor_name: str, data: PYJobInput) -> PYJobOutput:
+        if data.job_id:
+            raise HTTPException(
+                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
+                detail=f"Job id field is set but must not be: {data.job_id}"
+            )
+        data.job_id = generate_id()  # Generate processing job id
+
         if data.agent_type not in ['worker', 'server']:
             raise HTTPException(
                 status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                 detail=f"Unknown network agent with value: {data.agent_type}"
             )
+        workspace_db = await db_get_workspace(
+            workspace_id=data.workspace_id,
+            workspace_mets_path=data.path_to_mets
+        )
+        if not workspace_db:
+            raise HTTPException(
+                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
+                detail=f"Workspace with id: {data.workspace_id} or path: {data.path_to_mets} not found"
+            )
+
+        # The workspace is currently locked (being processed)
+        # TODO: Do the proper caching here, after the refactored code is working
```

Comment:
Objection by @MehmedGIT in today's call: Processing Server clients (like the Workflow Server) could decide to split up the job into several page ranges (esp. if they know there are multiple worker instances running in the back) and issue them concurrently. So instead of a boolean or string, we would need to synchronise over the actual set of pages of each workspace.

Reply:
> - check here whether canonical serialization of data is the same as being_processed. If so, allow the request, even though workspace_db is being_processed.

This is not achieving proper workspace locking when the same request is created with overlapping page ranges for the same workspace.

> we would need to synchronise over the actual set of pages of each workspace.

This is the way to achieve proper page range locking to prevent collisions on the processor server/worker level, however, also complex to implement. The more I think about it the more problematic it seems when considering error handling.

1) Ideal scenario example: 200 pages workspace, the Workflow Server will create 4 processing requests (50 pages per request) when knowing there are 4 workers of the specified processor. Then each request's `page_id` field will be in the form `start_page_id..end_page_id`. The `being_processed` boolean field will be replaced with a str field `locked_pages` in the form `start_page_id1..end_page_id1,...,start_page_id4..end_page_id4`. Then detecting overlapping single pages or page ranges is achievable to raise errors for the next coming requests. If the ranges don't overlap and match for the same processor, then the workspace would not be considered locked. For other processors, the workspace will still be considered locked. The unlocking of the workspace for pages will then happen based on the internal callbacks to the Processing Server from the Processor Server or Processing Worker when the execution finishes or fails. 

Problem: Still, in the case when either of the two fails to use a callback, say due to a crash, the workspace will be indefinitely locked. Considering a proper timeout for a server or worker is hard.

 2) Another point:
 Considering that `page_id` could be: 
- a single page - `PHYS0001`
- multiple pages separated by a comma - `PHYS0003,PHYS0004,PHYS0005`
- a page range... `PHYS0009..PHYS0015`

And now assuming a potential scenario where 3 processing requests with different `page_id` formats above are passed - the value of `locked_pages` for the 3 requests will be `PHYS0001_PHYS0003,PHYS0004,PHYS0005_PHYS0009..PHYS0015` assuming that the separator will be an underscore. Or just a string list with values `[PHYS0001][PHYS0003,PHYS0004,PHYS0005][PHYS0009..PHYS0015]`.

Potential problem: `locked_pages` is filled with a lot of values in case single page requests are passed for the entire workspace.

Reply:
> The unlocking of the workspace for pages will then happen based on the internal callbacks to the Processing Server from the Processor Server or Processing Worker when the execution finishes or fails.
> 
> Problem: Still, in the case when either of the two fails to use a callback, say due to a crash, the workspace will be indefinitely locked. Considering a proper timeout for a server or worker is hard.

So with the new model of shifting responsibility for job dependencies (i.e. workspace locking) from the external caller to the Processing Server and therefore using internal callbacks in addition to the requested callbacks, you are asking what if the internal callbacks fail? (BTW, you could also ask what happens if the negative acknowlegement to the queue fails, or the queue itself crashes.)

Well, we still have our database, don't we? Instead of adding to the DB only after unblocking and queuing the job, we should probably persist the request right away, but with your new `state=StateEnum.CACHED`. So we'll now have  cached and queued and running job entries. For running jobs, we agreed there is a need for some kind of timeout anyway, universally. It gets enforced by
- Processor Server model: the calling Processing Server or client CLI
- Processing Worker model: the worker instance (around `invoke_processor()`) or the processor itself (inside `process()`)

Whoever witnessed a timeout must then either resubmit or roll back upstream, i.e. push back on the queue.

The same could happen if anything goes wrong on the queue – reinstate from the DB or resubmit to the internal queue.

Reply:
> Why raise errors? IMO, locking is simply about blocking requests until the associated resources become free again.

Right. I was just considering the cases where double executions of the same processor with overlapping pages are submitted. That would mostly end in error, wouldn't it? Unless overwrite is already set for the next requests.

> Why make this contingent on the identity of the processor?

Because the workspace will be unlocked for the same processor requests that potentially can have different page ranges. The collisions for the same processor but different pages are handled by the METS server. However, if we get a different processor but the same pages, then the workspace is locked and that request goes into the waiting internal queue (CACHED). Consider cases when running processors cannot be done independently, i.e., the output of the first processor has to be used by the second processor. 

> IMO, if the other request concerns a disjunct set of pages, then there simply cannot be a dependency relation to the currently running processor(s)

Right, that means we must track `processor · set(pages) · fileGrp` (considering the info from the next paragraph). This becomes even harder to track (and potentially debug) what to submit directly in the RabbitMQ and what to queue internally (to cache).

> If anything, if we want to look more closely into dependencies, we should consider pairs set(pages) · fileGrp, where
    - a new request must block if its pages and input fileGrp is overlapping the locked pairs, and
    - a new job will lock its pages and output fileGrp while running
Thus, even requests on the same pages but regarding distinct fileGrps would stay independent.

Oh, right. We should also consider fileGrp in the mix of complexity...

> Not a problem: simply use the actual resolved list (or rather, set) of pages! Set operations (disjunction for adding, intersection for testing, difference for clearing) are clear and efficient.

This is of course simpler to manage although there will be more stress on the DB itself with reads/writes - which I guess is okay for now.

> So we'll now have cached and queued and running job entries. For running jobs, we agreed there is a need for some kind of timeout anyway, universally

Yes, we have some dummy timeout right now based on the submitted amount of pages (set to 200 by default multiplicated with a timeout value per page).  Which could potentially lead to timeout errors even when the output is correctly produced say due to slower processing. The amount of pages is still not part of the processing request. 

> BTW, IINM we still need to take care of atomicity in some places, i.e. use a mutex to query and change the internal queue...

Yes, there will be a race condition to the internal queue resource.

Reply:
> > IMO, if the other request concerns a disjunct set of pages, then there simply cannot be a dependency relation to the currently running processor(s)
> 
> Right, that means we must track `processor · set(pages) · fileGrp` (considering the info from the next paragraph). This becomes even harder to track (and potentially debug) what to submit directly in the RabbitMQ and what to queue internally (to cache).

Sorry to add to the complexity but we also need the parameterization of the processor in the mix, so `processor · parameters · set(pages) · fileGrp` 😬

Reply:
> I was just considering the cases where double executions of the same processor with overlapping pages are submitted. That would mostly end in error, wouldn't it? Unless overwrite is already set for the next requests.

Oh that. Yes, that's right. But for that we don't have to do anything (detecting overlaps etc) – this error will be coming from the processor itself (it already contains logic looking for conflicts in the output fileGrp / page list).

> > Why make this contingent on the identity of the processor?
> 
> Because the workspace will be unlocked for the same processor requests that potentially can have different page ranges. The collisions for the same processor but different pages are handled by the METS server. However, if we get a different processor but the same pages, then the workspace is locked and that request goes into the waiting internal queue (CACHED). Consider cases when running processors cannot be done independently, i.e., the output of the first processor has to be used by the second processor.

This sounds confusing to me. My [above analysis](https://github.com/OCR-D/core/pull/1069#discussion_r1259789028) still stands IMHO – I don't see any reason why we should look at the identity of the processor. (But I agree with @kba that if we do, then it would really have to be the combination of processor and parameters.) It's the just pages and the fileGrps that cause a dependency.

Regarding timeouts (using the actual number of pages in the Processor Server model, adding a timeout mechanism in the Processing Worker) – should we track that in a separate issue?

Reply:
> Oh that. Yes, that's right. But for that we don't have to do anything

Agree, we should not try to do early prevention of errors since it makes the implementation more complex on the Processing Server side.

> It's the just pages and the fileGrps that cause a dependency.

For the sake of keeping it simple - it should be just pages and fileGrp of a workspace then.

Reply:
> Regarding timeouts (using the actual number of pages in the Processor Server model, adding a timeout mechanism in the Processing Worker) – should we track that in a separate issue?

https://github.com/OCR-D/core/issues/1074

> Agree, we should not try to do early prevention of errors since it makes the implementation more complex on the Processing Server side.

Yes. If we want to do anticipation of workflow conflicts then it would be what currently is done in `ocrd process` via `ocrd.task_sequence.validate_tasks()` – checking that
- all processors exist
- all parameters are valid for all steps
- no output fileGrp exists prior to that step (on the data or the workflow itself), except when using `--overwrite`
- no input fileGrp will be missing for any step (because already on the data or generated earlier on the workflow)

That could be done in the Workflow Server statically – before sending requests.